pkg/webhooks/cluster_webhook.go

```diff
@@ -0,0 +1,204 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+
+	v1 "kurator.dev/kurator/pkg/apis/cluster/v1alpha1"
+)
+
+const (
+	defaultPodCIDR     = "192.168.0.0/16"
+	defaultServiceCIDR = "10.96.0.0/12"
+	defaultVpcCIDR     = "10.0.0.0/16"
+)
+
+var _ webhook.CustomDefaulter = &ClusterWebhook{}
+var _ webhook.CustomValidator = &ClusterWebhook{}
+
+type ClusterWebhook struct {
+	Client client.Reader
+}
+
+func (wh *ClusterWebhook) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(&v1.Cluster{}).
+		WithDefaulter(wh).
+		WithValidator(wh).
+		Complete()
+}
+
+func (wh *ClusterWebhook) Default(_ context.Context, obj runtime.Object) error {
+	in, ok := obj.(*v1.Cluster)
+	if !ok {
+		return apierrors.NewBadRequest(fmt.Sprintf("expected a Cluster but got a %T", obj))
+	}
+
+	if len(in.Spec.Network.PodCIDRs) == 0 {
+		in.Spec.Network.PodCIDRs = []string{defaultPodCIDR}
+	}
+
+	if len(in.Spec.Network.ServiceCIDRs) == 0 {
+		in.Spec.Network.ServiceCIDRs = []string{defaultServiceCIDR}
+	}
+
+	if in.Spec.Network.VPC.CIDRBlock == "" {
+		in.Spec.Network.VPC.CIDRBlock = defaultVpcCIDR
+	}
+
+	return nil
+}
+
+func (wh *ClusterWebhook) ValidateCreate(_ context.Context, obj runtime.Object) error {
+	in, ok := obj.(*v1.Cluster)
+	if !ok {
+		return apierrors.NewBadRequest(fmt.Sprintf("expected a Cluster but got a %T", obj))
+	}
+
+	return wh.validate(in)
+}
+
+func validateInfra(in *v1.Cluster) field.ErrorList {
+	var allErrs field.ErrorList
+	if in.Spec.InfraType == "" {
+		allErrs = append(allErrs, field.Required(field.NewPath("spec", "infraType"), "must be set"))
+	}
+
+	if in.Spec.Region == "" {
+		allErrs = append(allErrs, field.Required(field.NewPath("spec", "region"), "must be set"))
+	}
+
+	if in.Spec.Credential == nil || in.Spec.Credential.SecretRef == "" {
+		allErrs = append(allErrs, field.Required(field.NewPath("spec", "credential", "secretRef"), "must be set"))
+	}
+
+	return allErrs
+}
+
+func validateVersion(in *v1.Cluster) field.ErrorList {
+	var allErrs field.ErrorList
+	if in.Spec.Version == "" {
+		allErrs = append(allErrs, field.Required(field.NewPath("spec", "version"), "must be set"))
+	} else {
+		if !version.KubeSemver.MatchString(in.Spec.Version) {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("spec", "version"), in.Spec.Version, "must be a valid Kubernetes version"))
+		}
+	}
+
+	return allErrs
+}
+
+func validateNetwork(network v1.NetworkConfig, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	allErrs = append(allErrs, validateCIDR(network.VPC.CIDRBlock, fldPath.Child("vpc", "cidrBlock"))...)
+	allErrs = append(allErrs, validateCIDRBlocks(network.PodCIDRs, fldPath.Child("podCIDRs"))...)
+	allErrs = append(allErrs, validateCIDRBlocks(network.ServiceCIDRs, fldPath.Child("serviceCIDRs"))...)
+
+	return allErrs
+}
+
+func validateCIDR(cidr string, fldPath *field.Path) field.ErrorList {
```

Comment:
duplicate with validateCIDRBlocks？

Reply:
difference between `[]string` and `string`